src/client/sandbox/xhr.js

```diff
@@ -0,0 +1,182 @@
+    constructor (sandbox) {
+        super(sandbox);
+
+        this.XHR_COMPLETED_EVENT = 'hammerhead|event|xhr-completed';
+        this.XHR_ERROR_EVENT     = 'hammerhead|event|xhr-error';
+        this.XHR_SEND_EVENT      = 'hammerhead|event|xhr-send';
+
+        var xhr = new nativeMethods.XMLHttpRequest();
+
+        this.corsSupported = typeof xhr.withCredentials !== 'undefined';
+    }
+
+    static createNativeXHR () {
+        var xhr = new window.XMLHttpRequest();
+
+        xhr[NATIVE_BEHAVIOR] = true;
+
+        return xhr;
+    }
+
+    attach (window) {
+        super.attach(window);
+
+        var xhrSandbox = this;
+
+        window.XMLHttpRequest.prototype.abort = function () {
+            nativeMethods.xmlHttpRequestAbort.call(this);
+            if (!this[NATIVE_BEHAVIOR]) {
+                xhrSandbox.emit(xhrSandbox.XHR_ERROR_EVENT, {
+                    err: new Error('XHR aborted'),
+                    xhr: this
+                });
+            }
+        };
+
+        // NOTE: Redirect all requests to the Hammerhead proxy and ensure that requests don't
+        // violate Same Origin Policy.
+        window.XMLHttpRequest.prototype.open = function (method, url, async, user, password) {
+            // NOTE: Emulate CORS, so that 3rd party libs (e.g. jQuery) allow requests with the proxy host as well as
+            // the destination page host.
+            if (!xhrSandbox.corsSupported)
+                this.withCredentials = false;
+
+            if (!this[NATIVE_BEHAVIOR]) {
+                var xhr = this;
+
+                try {
+                    url = getProxyUrl(url);
+                }
+                catch (err) {
+                    xhrSandbox.emit(xhrSandbox.XHR_ERROR_EVENT, { err, xhr });
+
+                    return;
+                }
+            }
+
+            // NOTE: The 'async' argument is true by default. However, when the 'async' argument is set to â€˜undefinedâ€™,
+            // a browser (Chrome, FireFox) sets it to 'false', and a request becomes synchronous (B238528).
+            if (arguments.length === 2)
+                nativeMethods.xmlHttpRequestOpen.call(this, method, url);
+            else
+                nativeMethods.xmlHttpRequestOpen.call(this, method, url, async, user, password);
+        };
+
+        window.XMLHttpRequest.prototype.send = function () {
+            if (!this[NATIVE_BEHAVIOR]) {
+                var xhr = this;
+
+                xhrSandbox.emit(xhrSandbox.XHR_SEND_EVENT, { xhr });
+
+                var orscHandler = () => {
+                    if (this.readyState === 4)
+                        xhrSandbox.emit(xhrSandbox.XHR_COMPLETED_EVENT, { xhr });
+                };
+
+                // NOTE: If we're using the sync mode or the response is in cache and the object has been retrieved
+                // directly (IE6 & IE7), we need to raise the callback manually.
+                if (this.readyState === 4)

```

Comment:
IE6, IE7 is not supported browsers.
Do we need this statement now? If it is right then we should update the comment.